doc/DESIGN.md

```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
 The reason behind this is, that the above functions – as their name
 suggest – only retrieve values. The returned value must not be modified
 directly.
+
+## Design Guidelines Checklist
+
+On potential changes of the API/ABI as detected by
+[`elektra-icheck`](https://build.libelektra.org/job/elektra-icheck),
+please make sure the API has been reviewed according to the
+following 2 checklists:
+
+## Checklist for overall API
+
+### Naming
+
+- [ ] Inconsistent naming of functions or variables
+
+### Consistency
+
+- [ ] Consistent naming schemes
+- [ ] Similar things are named similarly
+- [ ] Different things are named differently
+- [ ] The order of arguments should be consistent across similar functions
+- [ ] Code is formatted according to the [style guidelines](/doc/CODING.md#coding-style)
+
+### Structural Clarity
+
+- [ ] Functions with similar functionality are grouped into the same namespace
+
+### Compatibility
+
+- [ ] All Bindings have been updated to reflect the new API and work properly
+
+### Extensibility
+
+- [ ] New API is easily extensible with additional functionality
```

Comment:
Unclear how this can be checked.

Reply:
Yes this was unclear to me as well. You mentioned it in our email conversation - so I put it up here for visibility. I haven't really found any good heuristics for this yet.

Reply:
Yes, being futureproof is a difficult topic.

One part of it is to make sure that we are generic, e.g. general flags instead of specific locking flags. But being too generic also is problematic as people might not think of that they can lock a key via setting flags (or remove a key via lookup).

One other part of it is to have as little assumptions as possible on what users might want to do.

Did you find some papers about the topic?

Reply:
I haven't really found anything that can be easily distilled into a checklist. I have been doing further research into this area, so I hope I can can come up with something soon.